### PR TITLE
feat(gossipsub): use signed peer records on handle prune

### PR DESCRIPTION
**Pull request by dapplion**
## Description

Gossipsub's PeerExchange functionality suggests peers to connect to upon pruning a node from the mesh. Currently, we cannot import _sending_ signed peer records to other nodes because we don't have a peer store (yet). This patch implements handling incoming signed peer records from other nodes in case we get pruned from the mesh.

Related https://github.com/libp2p/rust-libp2p/issues/2398.

## Notes

N/A

## Links to any relevant issues

- https://github.com/libp2p/rust-libp2p/issues/2398

## Open Questions

N/A

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
  - Could not find any docs affected, please indicate otherwise 
- [x] I have added tests that prove my fix is effective or that my feature works
  - A test covers dialing peers from signed_peer_records, but does not assert that addresses is correct, see

https://github.com/libp2p/rust-libp2p/blob/d1d468c13c8df04ef8b52566c58961b24652515c/protocols/gossipsub/src/behaviour/tests.rs#L1896-L1897

- [ ] A changelog entry has been made in the appropriate crates
  - Where should I append the entry? Add a new `# 0.45.0` section?

### CODE REVIEW

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour/tests.rs:1910]: I think it would be good to assert the presence of the addresses too!

**Review by thomaseizinger — COMMENTED** [review]
Thanks! This is definitely an improvement by itself.

**Comment by thomaseizinger** [comment]
Regarding changelog: Given that this is backwards-compatible, can you can add an entry to the `0.44.1` section.

**Comment by dapplion** [comment]
cargo-semver-checks CI step errors with
```
    Finished dev [unoptimized + debuginfo] target(s) in 38.99s
     Parsing libp2p-core v0.39.0 (baseline, cached)
Error: version mismatch between current (v24) and baseline (v23) format versions
Error: Process completed with exit code 1.
```
Any pointers on what could be the root issue? https://github.com/libp2p/rust-libp2p/actions/runs/4380836480/jobs/7668318386#step:14:409

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (dapplion) [reply protocols/gossipsub/src/behaviour/tests.rs:1910]: Any suggestion on how to retrieve the addresses from DialOpts struct? 

https://github.com/libp2p/rust-libp2p/blob/3f5dd631aef6592fbcb1439bdd311ae9f81be684/protocols/gossipsub/src/behaviour/tests.rs#L1896-L1897

Or should I make this fn public?

https://github.com/libp2p/rust-libp2p/blob/d81f9476f72a6eb1d4bc683ad04b3fdf866cfc12/swarm/src/dial_opts.rs#L130-L132

**Comment by thomaseizinger** [comment]
> cargo-semver-checks CI step errors with
> 
> ```
>     Finished dev [unoptimized + debuginfo] target(s) in 38.99s
>      Parsing libp2p-core v0.39.0 (baseline, cached)
> Error: version mismatch between current (v24) and baseline (v23) format versions
> Error: Process completed with exit code 1.
> ```
> 
> Any pointers on what could be the root issue? [libp2p/rust-libp2p/actions/runs/4380836480/jobs/7668318386#step:14:409](https://github.com/libp2p/rust-libp2p/actions/runs/4380836480/jobs/7668318386#step:14:409)

Uhm this is likely because the caches are stale. I'll delete them.

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour/tests.rs:1910]: I am not really a fan of the style of the tests in gossipsub anyway. We should only be testing behaviours by running them in a `Swarm` and not call these lifecycle functions manually.

We now have a new crate, `libp2p-swarm-test` which makes it easy to set up swarms for testing.

How about we do the following:

- Implement "sending" of signed peer records by keeping a hashmap of records in the behaviour where we perform a lookup upon prune
- At the moment, we don't have a way of adding records to this hashmap so in reality, there won't ever be any
- Within the test, we can add a record (by accessing the field directly)
- We can then write a test using `libp2p-swarm-test` that ends up sending a `Prune` to a node which will then contain the `SignedPeerRecord`
- We can then assert that the `Swarm` establishes a connection to the node specified in the prune message

What do you think?

Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour/tests.rs:1911]: I am not really a fan of the style of the tests in gossipsub anyway. We should only be testing behaviours by running them in a `Swarm` and not call these lifecycle functions manually.

We now have a new crate, `libp2p-swarm-test` which makes it easy to set up swarms for testing.

How about we do the following:

- Implement "sending" of signed peer records by keeping a hashmap of records in the behaviour where we perform a lookup upon prune
- At the moment, we don't have a way of adding records to this hashmap so in reality, there won't ever be any
- Within the test, we can add a record (by accessing the field directly)
- We can then write a test using `libp2p-swarm-test` that ends up sending a `Prune` to a node which will then contain the `SignedPeerRecord`
- We can then assert that the `Swarm` establishes a connection to the node specified in the prune message

What do you think?

**Comment by mxinden** [comment]
Let me know if you need a review from my end. Also adding @divagant-martian and @AgeManning for visibility and, if time permits, a review.

File: protocols/gossipsub/src/behaviour.rs
Comment (AgeManning) [inline protocols/gossipsub/src/behaviour.rs:1644]: I know this should have been done earlier, probably on line 1617. 

I think we probably don't want to try to connect to all the peers, we probably only want to try and get up to `mesh_n_high`. I'd suggest we set `n` on line 1617 to be the min of `self.config.prune_peers()` or `self.mesh.get(&topic_hash).map(|x| x.len()).unwrap_or(0)` or something. Which may mean calculating this in `handle_prune()` because we don't have the topic hash here.

**Review by AgeManning — COMMENTED** [review]
Yeah nice. Looks good. Just a small comment

File: protocols/gossipsub/src/behaviour.rs
Comment (divagant-martian) [inline protocols/gossipsub/src/behaviour.rs]: this only ignores peer exchange for the current peer in the list. I think we might want to halt connecting to any peer received on the exchange if this is invalid

File: protocols/gossipsub/src/behaviour.rs
Comment (divagant-martian) [inline protocols/gossipsub/src/behaviour.rs:1625]: I can't add a comment higher up in the code, but you probably want to remove lines 1618 to 1622
```rust
// Ignore peerInfo with no ID
//
//TODO: Once signed records are spec'd: Can we use peerInfo without any IDs if they have a
// signed peer record?
px.retain(|p| p.peer_id.is_some());
```

File: protocols/gossipsub/src/behaviour.rs
Comment (divagant-martian) [reply protocols/gossipsub/src/behaviour.rs:1644]: Agree with the needed change in general. 

Note that `prune_peers` according to our docs configure the number of peers we _send_ to others, not the number of peers we connect to. Either code or docs need to be adjusted. But using the current docs, this should probably not be used here (or in `handle_prune`)

It's also not clear to me if handling the prune is the best place to connect to those peers. We already graft peers in the heartbeat when below `mesh_n_low`. We might want to connect to these peers to accelerate the process, or we might want to prioritize already connected peers since they are somewhat more trusted. In any case, both `mesh_n_high` and `mesh_n_low` should probably be used to select the number of peers from this set we want to try.

**Review by divagant-martian — COMMENTED** [review]
partial review since it's getting late in my timezone

File: protocols/gossipsub/src/behaviour.rs
Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour.rs:1644]: PeerExchange is meant to give you more peers in case you are only connected to a few, right?

Perhaps it would be better to just add those to some kind of list and let the grafting logic upon heartbeats do the connecting?

For the case where we don't have many peers, that would otherwise "fail" in that we can't connect to as many peers as we'd like.

Comment (dapplion) [reply protocols/gossipsub/src/behaviour.rs:1644]: Those are good points, and path forward is opinionated. Maintainers and @divagant-martian @AgeManning could you comment on what makes more sense?

IMO connecting eagerly instead of accumulating keep it simpler + limit only up to mesh needs. For reference Go impl eagerly schedules all peer exchanges to the connect channel (with max cap MaxPendingConnections)

File: protocols/gossipsub/src/behaviour.rs
Comment (dapplion) [reply protocols/gossipsub/src/behaviour.rs]: Then, should it collect all decoded records first, then iter and connect to all?

File: protocols/gossipsub/src/behaviour.rs
Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour.rs:1644]: I've not touched gossipsub much so I'll leave it to @mxinden and @AgeManning to make a decision here.

> For reference Go impl eagerly schedules all peer exchanges to the connect channel (with max cap MaxPendingConnections)

That is an interesting data point.

Comment (AgeManning) [reply protocols/gossipsub/src/behaviour.rs:1644]: Yeah, now that I think about this, there's not that many behaviours that dial peers. 

In rust-libp2p, when we dial a peer, we not only connect to this behaviour, but we connect to all behaviours that the peer supports. In Lighthouse, we are pretty strict about which peers we dial. We have limits on outbound vs inbound connections, banned ips, scoring etc and so have very stringent dialing conditions. It would actually kind of cause havoc for us if gossipsub was out and about dialing random peers that are just feed to us. In principle perhaps they could form eclipse attacks but feeding us malicious peers constantly. 

I think there are two ways forward and I'd be curious to hear everyone's opinion on them, because I dont have a strong preference:
1) We keep it as is an dial eagerly provided we have the need, i.e that `mesh_high` > `connected_peers[topic_hash].len()` (we could potentially use `mesh_n` but probably high is better). This is somewhat safe, because px is disabled by default in the config and turning it on should be done at one's own risk. 
2) We create a new event type. Gossipsub can propagate upwards potential peers to the swarm and the user can decide whether to dial or not. 

After writing these out, I think 1 is probably the easiest as I realize its off by default and can be opted in to.

Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour.rs:1644]: (2) is also touched on in https://github.com/libp2p/rust-libp2p/issues/2680. At the moment, there is no way for a behaviour to in a standard way emit an event that it learned addresses about a peer. This makes it difficult to build a general address-book style behaviour.

I do think it would be a good idea but the details would need to be fleshed out. One idea would be to immediately re-inject this information into the behaviour, therefore effectively distributing it across all behaviours. One of them could then be a (persistent) address book that uses the provided addresses to enhance future dial attempts.

File: protocols/gossipsub/src/behaviour.rs
Comment (divagant-martian) [inline protocols/gossipsub/src/behaviour.rs]: this is likely to panic on ~~release~~ debug

Also, I think you want `mesh_n_high` instead of `mesh_n_low` ? Both are fine depending on what we want, @AgeManning suggested the high one which would be interpreted as capacity more than need

File: protocols/gossipsub/src/behaviour.rs
Comment (divagant-martian) [inline protocols/gossipsub/src/behaviour.rs]: another line no longer in sync with code

File: protocols/gossipsub/src/behaviour.rs
Comment (divagant-martian) [reply protocols/gossipsub/src/behaviour.rs:1644]: For the sake of unblocking this PR we can go with eager connection, since as Age mentioned, this needs to be explicitly turned on.

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (divagant-martian) [inline protocols/gossipsub/src/behaviour/tests.rs:1902]: likely to fail?

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (divagant-martian) [reply protocols/gossipsub/src/behaviour/tests.rs:1910]: going this way seems reasonable to me

**Review by divagant-martian — CHANGES_REQUESTED** [review]
Except for a risk pf panic and test updates I think this is close to ready

File: protocols/gossipsub/src/behaviour.rs
Comment (dapplion) [reply protocols/gossipsub/src/behaviour.rs:1644]: @AgeManning Should `do_px` be used to not consume px by default? Or add a new config param like `use_px` to split provision and consumption of px

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (dapplion) [reply protocols/gossipsub/src/behaviour/tests.rs:1910]: That makes tho if all tests in this package are written with a consistent style, should this PR stick to that style? And then refactor them into using swarm-test in another PR

**Comment by dapplion** [comment]
Rebased to latest master to fix conflicts. Since the discussion is getting long I'll link the remaining points:
- Ok to eagerly connect for now, but must be disabled by default. With which config param? https://github.com/libp2p/rust-libp2p/pull/3579#discussion_r1152726969
- Test using libp2p-swarm-test to assert dialed addresses https://github.com/libp2p/rust-libp2p/pull/3579#discussion_r1152739665

File: protocols/gossipsub/src/behaviour.rs
Comment (AgeManning) [reply protocols/gossipsub/src/behaviour.rs:1644]: I think we should have a more explicit config. I think we should opt for `enable_peer_exchange()` which enables both sending and receiving. I dont think users really need granularity over one or the other (we can add it later if its requested). 

I'd suggest renaming `do_px` to `enable_peer_exchange`. Have it off my default.

Comment (divagant-martian) [reply protocols/gossipsub/src/behaviour.rs:1644]: @dapplion to achieve this you technically need to change nothing. This is already the case. Age's suggestion is a nice to have, but would change the public api, so the version should be updated accordingly. I'm fine with doing or not doing this change

File: protocols/gossipsub/src/behaviour/tests.rs
Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour/tests.rs:1910]: > That makes tho if all tests in this package are written with a consistent style, should this PR stick to that style? And then refactor them into using swarm-test in another PR

Eventually yes but that is a massive effort. I don't mind mixing styles in a transition period if we get a better, more feature-oriented test in exchange.

File: protocols/gossipsub/src/behaviour.rs
Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour.rs:1644]: We can do it in a non-breaking way by adding `enable_peer_exchange` on `Config` and deprecating the `do_px` function. Internally, the field can be renamed without issues.

**Comment by dapplion** [comment]
Closing PR for now since I have no capacity to address it. If there's a real need for this feature I'm happy to prioritise it again, but from offline discussions it seems like it is not the case. Sorry for polluting your PR inbox for a while!

**Pull request closed without merging**
